readme.md
# AFScroll (Animation Frame Scroll)

The script aligns page scroll update with animation frame.  
It allows to perform js based DOM manipulation related to page scroll just in time, not too late.

[Demo](https://dobrapyra.github.io/af-scroll/)

**Menu**

* [Quick start](#quick-start)
* [Features](#features)
* [Disadvantages](#disadvantages)
* [Why you may need AFScroll?](#why-you-may-need-afscroll?)
* [Options](#options)
* [Methods](#methods)
* [Events](#Events)

## Quick start

### Install

```bash
# install af-scroll npm package
npm install --save af-scroll
```

### Initialization

```js
// import create function
import createAFScroll from 'af-scroll';

// call to initialize
createAFScroll();
```

## Imports

The library build is delivered in two formats: `esmodule` and `commonjs`

```js
// default import (esmodule)
import createAFScroll from 'af-scroll';

// explicit import (esmodule)
import createAFScroll from 'af-scroll/dist/esmodule';

// explicit import (commonjs)
const createAFScroll = require('af-scroll/dist/commonjs');
```

## Features

* alignment of js on scroll updates with the visual scroll update
* on scroll updates throttled by requestAnimationFrame
* native scroll triggers (native scrollbar, mouse wheel, keyboard, touch)
* scroll smoothing (optional)
* scroll locking (optional)

## Disadvantages

* additional wrapping element in DOM 
* missing support for native scroll to results of find on page

## Why you may need AFScroll?

If you want to manipulate DOM elements by js and set them based on the scroll position.

### What happens when you use default scroll:

* user scrolls a page
* browser renders new scroll position
* **user sees new scroll position**
* the native on scroll event is triggered
* js updates for current scroll position  
(any custom DOM manipulations by js)
* **user sees DOM elements updated by js**

### What happens when you use AFScroll:

* user scrolls a page
* browser renders new scroll position
* user not sees new scroll position
* the native on scroll event is triggered
* js updates for current scroll position  
(update visible scroll position & any custom DOM manipulations by js)
* **user sees new scroll position & DOM elements updated by js in the same time**

## Options

```js
createAFScroll({
  smoothForce: 0.8,
  smoothLimit: 0.2,
  scrollEl: null,
  className: 'afScroll',
  wrapExclude: 'script, link',
  autoHeight: 12,
  onUpdate: () => {},
  onComplete: () => {},
});
```

| Option | Type | Default | Description |
| --- | --- | --- | --- |
| `smoothForce` | Number | `0.8` | smoothing force in range 0.0 - 1.0<br />`0`: no smooth<br />`1`: no scroll |
| `smoothLimit` | Number | `0.2` | min diff between current and target value to keep smooth loop |
| `scrollEl` | Element | `null` | static scroll wrapper element<sup>1</sup> |
| `className` | String | `"afScroll"` | css class name of scroll wrapper element<sup>2</sup> |
| `wrapExclude` | String | `"script, link"` | css selector to exclude from wrapping<sup>2</sup> |
| `autoHeight` | Number | `12` | height checking period in frames<br />`1`: each frame<br />`0`: disabled<br />`12`: once per each 12 frames |
| `onUpdate` | Function | `() => {}` | callback function triggered on scroll update |
| `onComplete` | Function | `() => {}` | callback function triggered after smooth loop stopped |

1. prevents DOM modification by auto created wrapper, passed `scrollEl` will be used instead of it
2. ignored if `scrollEl` passed

## Methods

```js
// example of using methods
const afScroll = createAFScroll(/* optional options */);
afScroll.scrollTo(0); // scroll to top
afScroll.scrollTo(0, true); // scroll to top forced
afScroll.free(); // free forced scroll
afScroll.lock(); // lock scroll
afScroll.unlock(); // unlock scroll
afScroll.destroy(); // destroy af-scroll
afScroll.init(); // reinit af-scroll
```

| Method | Arguments | Return | Description |
| --- | --- | --- | --- |
| `init()` | none | undefined | reinitialize the script after destroy |
| `scrollTo(scroll, force)` | `scroll` {Number} target scroll value in px unit<br />`force` {Boolean} prevents user from changing a scroll target until it completes | undefined | scroll to `scroll` value |
| `free()` | none | undefined | free forced scroll |
| `lock()` | none | undefined | lock scroll |
| `unlock()` | none | undefined | unlock scroll |
| `destroy()` | none | undefined | destroy the script and restore original html |

## Events

```js
createAFScroll({
  onUpdate: (scroll) => {
    // custom behavior
  },
  onComplete: (scroll) => {
    // custom behavior
  },
});
```

| Event | Arguments | When triggered |
| --- | --- | --- |
| `onUpdate(scroll)` | `scroll` {Number} scroll value in px unit | in each animation frame of move to target scroll |
| `onComplete(scroll)` | `scroll` {Number} scroll value in px unit | after achieve target scroll - smooth loop stopped |

changelog.md
# Changelog

## 1.1.1 (26.03.2023)

Modified:
* Internal micro optimalization of `scrollTo`.
* Changed default export to `esmodule`.

---

## 1.1.0 (25.03.2023)

Added:
* Optional parameter `force` in `scrollTo(scroll, force)` method.
* Method `free()` to free forced scroll.

---

## 1.0.0 (12.03.2023)

Initial release.

---


src/index.js
import { each, lerp, style } from './helpers';

/**
 * AFScroll (Animation Frame Scroll)
 * @param {Object} options general options
 * @property {Number} smoothForce smoothing force in range 0.0 - 1.0, 0: no smooth, 1: no scroll
 * @property {Number} smoothLimit min diff between current and target value to keep smooth loop
 * @property {Element} scrollEl static scroll wrapper element
 * @property {String} className css class name of scroll wrapper element
 * @property {String} wrapExclude css selector to exclude from wrapping
 * @property {Number} autoHeight height checking period in frames, 1: each frame, 0: disabled
 * @property {Function} onUpdate callback function triggered on scroll update
 * @property {Function} onComplete callback function triggered after smooth loop stopped
 */
export default function createAFScroll({
  smoothForce: smoothForceArg = 0.8,
  smoothLimit: smoothLimitArg = 0.2,
  scrollEl: scrollElArg = null,
  className: classNameArg = 'afScroll',
  wrapExclude: wrapExcludeArg = 'script, link',
  autoHeight: autoHeightArg = 12,
  onUpdate: onUpdateArg = () => {},
  onComplete: onCompleteArg = () => {}
} = {}) {
  const smoothFactor = 1 - smoothForceArg;
  const smoothLimit = smoothLimitArg;
  const staticScrollEl = scrollElArg;
  const className = classNameArg;
  const wrapExclude = wrapExcludeArg;
  const autoHeight = autoHeightArg;
  const onUpdate = onUpdateArg;
  const onComplete = onCompleteArg;

  let targetScroll = 0;
  let lastScroll = 0;
  let forcedScroll = null;
  let lockedScroll = null;
  let smoothRaf = null;
  let heightRaf = null;

  const bodyEl = document.getElementsByTagName('body')[0];
  let scrollEl = null;
  let autoHeightFrame = 0;
  let lastHeight = null;

  /**
   * create scroll wrapper element
   */
  function createScroll() {
    scrollEl = staticScrollEl !== null
      ? staticScrollEl
      : document.createElement('div');
    style(scrollEl, {
      position: 'fixed',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
      overflow: 'hidden',
    });

    if (staticScrollEl !== null) return;

    scrollEl.setAttribute('class', className);

    const childrenArr = [];
    each(bodyEl.children, childEl => {
      if (childEl === scrollEl || childEl.matches(wrapExclude)) return true;
      childrenArr.push(childEl);
    });
    each(childrenArr, childEl => {
      scrollEl.appendChild(childEl);
    });
    bodyEl.insertBefore(scrollEl, bodyEl.children[0]);
  }

  /**
   * remove scroll wrapper element
   */
  function removeScroll() {
    style(bodyEl, { height: '' });
    autoHeightFrame = 0;
    lastHeight = null;

    if (staticScrollEl !== null) {
      style(scrollEl, {
        position: '',
        top: '',
        left: '',
        width: '',
        height: '',
        overflow: '',
      });
      scrollEl = null;

      return;
    }

    const childrenArr = [];
    each(scrollEl.children, childEl => {
      childrenArr.push(childEl);
    });
    each(childrenArr, childEl => {
      bodyEl.insertBefore(childEl, scrollEl);
    });
    bodyEl.removeChild(scrollEl);
    scrollEl = null;
  }

  /**
   * setNativeScroll
   * @param {Number} scroll scrollTop value
   */
  function setNativeScroll(scroll) {
    document.body.scrollTop = document.documentElement.scrollTop = scroll;
  }

  /**
   * scrollTo
   * @public
   * @param {Number} scroll scrollTop value
   * @param {Boolean} force prevents user from changing a scroll target
   */
  function scrollTo(scroll, force = false) {
    if (force) forcedScroll = scroll;

    setNativeScroll(scroll);
  }

  /**
   * free forced scroll
   * @public
   */
  function free() {
    forcedScroll = null;
  }

  /**
   * lock scroll
   * @public
   */
  function lock() {
    lockedScroll = lastScroll;
  }

  /**
   * unlock scroll
   * @public
   */
  function unlock() {
    lockedScroll = null;
  }

  function updateScroll(scroll) {
    lastScroll = scroll;
    scrollEl.scrollTop = scroll;
    onUpdate(scroll);
  }

  function smoothUpdate() {
    if (lockedScroll !== null) return;

    if (Math.abs(targetScroll - lastScroll) < smoothLimit) {
      updateScroll(targetScroll);
      free();
      onComplete(targetScroll);
      return;
    }

    updateScroll(lerp(lastScroll, targetScroll, smoothFactor));

    smoothRaf = requestAnimationFrame(smoothUpdate);
  }

  function onScroll() {
    if (lockedScroll !== null) setNativeScroll(lockedScroll);
    else if (forcedScroll !== null) setNativeScroll(forcedScroll);

    targetScroll = window.scrollY;

    cancelAnimationFrame(smoothRaf);
    smoothRaf = requestAnimationFrame(smoothUpdate);
  }

  function breakSmoothLoop() {
    cancelAnimationFrame(smoothRaf);
  }

  function autoHeightUpdate() {
    if (++autoHeightFrame >= autoHeight) {
      autoHeightFrame = 0;
      updateHeight();
    }

    heightRaf = requestAnimationFrame(autoHeightUpdate);
  }

  function startAutoHeight() {
    if (autoHeight === 0 || autoHeight === false) return;

    heightRaf = requestAnimationFrame(autoHeightUpdate);
  }

  function stopAutoHeight() {
    cancelAnimationFrame(heightRaf);
  }

  function updateHeight() {
    const scrollHeight = scrollEl.scrollHeight;
    if (scrollHeight === lastHeight) return;

    lastHeight = scrollHeight;
    style(bodyEl, {
      height: `${scrollHeight}px`,
    });
  }

  function onResize() {
    updateHeight();
    updateHeight(); // double to fix some size issues
  }

  function bindEvents() {
    window.addEventListener('scroll', onScroll);
    window.addEventListener('resize', onResize);
  }

  function unbindEvents() {
    window.removeEventListener('scroll', onScroll);
    window.removeEventListener('resize', onResize);
  }

  /**
   * initialize AFScroll
   * @public
   */
  function init() {
    if (scrollEl !== null) return;

    targetScroll = window.scrollY;
    createScroll();
    onResize();

    setNativeScroll(targetScroll);
    updateScroll(targetScroll);

    bindEvents();
    startAutoHeight();
  }

  /**
   * destroy AFScroll
   * @public
   */
  function destroy() {
    if (scrollEl === null) return;

    breakSmoothLoop();
    stopAutoHeight();
    unbindEvents();
    removeScroll();

    forcedScroll = null;
    lockedScroll = null;
  }

  init();

  return {
    init,
    scrollTo,
    free,
    lock,
    unlock,
    destroy,
  };
}
